wurst/systems/trade/TradeShipSupervisor.wurst

```diff
@@ -0,0 +1,114 @@
+
+public enum TradeShipSupervisorState
+    MOVING
+    STOPPED
+    EXITED
+
+public class TradeShipSupervisor
+    static let ships = new HashMap<unit, thistype>()
+    static let whenEnterRegion = CreateTrigger()
+    ..registerEnterRegion(TradeShipStop.tradeStops, null)
+    ..addAction(function TradeShipSupervisor.handleEnterRegion)
+
+    unit ship
+    TradeShipStop firstStop
+    TradeShipStop nextStop
+    CallbackPeriodic callback
+    var state = TradeShipSupervisorState.STOPPED
+    let listeners = new LinkedList<TradeShipListener>()
+    var numForksReached = 0
+
+    static function handleEnterRegion()
+        let enteringUnit = GetEnteringUnit()
+        if ships.has(enteringUnit)
+            ships.get(enteringUnit).processEnterRegion()
+            
+    construct(TradeShipStop firstStop, int shipId)
+        ship = createUnit(players[PLAYER_NEUTRAL_PASSIVE], shipId, firstStop.getPos(), angle(0))
+        nextStop = firstStop.getNext()
+        ships.put(ship, this)
+        moveToNextStop()
+
+    function getState() returns TradeShipSupervisorState
+        return state
+
+    function addListener(TradeShipListener listener)
+        listeners.add(listener)
+
+    function triggerListeners()
+        listeners.forEach(listener -> listener.run(this))
+
+    function moveToNextStop()
+        if state == TradeShipSupervisorState.MOVING
+            return
+
+        state = TradeShipSupervisorState.MOVING
+        ship.issuePointOrder("move", nextStop.getPos())
+        
+        if callback != null
+            destroy callback
+            callback = null
+
+        callback = doPeriodically(1.0) cb ->
+            ship.issuePointOrder("move", nextStop.getPos())
+
+    function pause(real duration)
+        if state == TradeShipSupervisorState.STOPPED
+            return
+
+        if callback != null
+            destroy callback
+            callback = null
+            
+
+        state = TradeShipSupervisorState.STOPPED
+        doAfter(duration, -> moveToNextStop())
+
+    function removeShip()
+        ships.remove(ship)
+        ship.remove()
+        state = TradeShipSupervisorState.EXITED
+        ship = null
+
+    function setNextStop()
+        let hasFork = nextStop.hasFork()
+        if hasFork and numForksReached > 3
+            nextStop = nextStop.getFork()
+        else
+            numForksReached += hasFork.toInt()
+            nextStop = nextStop.getNext()
+
+    private function processEnterRegion()
+        if not nextStop.isInStopRange(ship)
+            return
+
+        if nextStop.isLastStop()
+            removeShip()
+        else
+            pause(nextStop.getPauseTime())
+            setNextStop()
+
+        triggerListeners()
+
+    ondestroy
+        listeners.forEach(listener -> destroy listener)
+        destroy listeners
+
+        if callback != null
+            destroy callback
+            callback = null
+            
+
+        if ship != null
```

Comment:
When could this happen in a proper game? Shouldn't the boat finish and be removed before destruction? If this is the case and this is a safety check, then print an error message here. Same for the callback.